### ISSUE TITLE
Carlini Wagner attack does not preturb a lot of the samples

### ISSUE THREAD

**Issue by anujstam** [IBM/adversarial-robustness-toolbox#352]
Hi,
I'm comparing two models for robustness and I'm attacking both with the Carlini Wagner L2 attack. I've noticed that the attack often returns the original sample back as the perturbed sample (IE, no change, L2 distance of zero, and no change of the confidence in the prediction by the model). The samples have fairly high confidence values to begin with so it is not a matter of the model being uncertain.

Across around 2000 samples, only 1000 were able to be altered for both the models. Any idea why this might be the case or how I can force the algorithm to run until I get some kind of altered sample?

Additionally, one of my networks is slightly unconventional, as it is a binary classifier, so in order to fit with the current requirements of ART, I had to manually convert it into a 2 output softmax, with fixed weights. The end model is exactly equivalent to the original, but I do not know if that may cause any problems. Nevertheless the first network is a normal 2 output CNN model, and I have the same issue there as well.

i am using the default parameters for the attack but I've set targeted=true and provided the adversarial classes as a vector of onehot encoded classes. The classifier is a keras classifier and use_logits is false. All the other parameters are default.

The dataset in question is the Asirra cats and dogs dataset : https://www.kaggle.com/c/dogs-vs-cats

Thank you for your time!

**Comment by mathsinn** [comment]
Hi @anujstam , thank you for using ART!

The Carlini Wagner attack will return the unmodified original input when the search for an adversarial example did not succeed. 

The first thing I'd try to make the attack succeed on a larger portion of samples is to increase the `initial_const` parameter by a couple of orders of magnitude. This will relax the attack objective to find adversarial examples with small L2 norm. You could also increase `max_iter`, e.g. to 100 or 1,000, but note this will increase computational costs.

Having said that, note that the Carlini Wagner attack is designed to operate on logits. If you could tweak your classifier to output logits instead of softmax, that should improve the overall attack performance.

Hope this helps - please let us know!

**Comment by anujstam** [comment]
Thank you! Will try and update ASAP

**Comment by beat-buesser** [comment]
Closing for now.
